### ISSUE TITLE
Handler for TestRunStarted event gets triggered after all the tests finish in case of parallel execution of tests

### ISSUE THREAD

**Issue by MazurK239** [cucumber/cucumber-jvm#2914]
### 👓 What did you see?

I registered a plugin that catches `TestRunStarted` and `TestRunFinished` events to do some set-up / tear-down logic.
This plugin extends `EventListener` and overrides method `setEventPublisher`.

When the tests are run sequentially (both via CLI and JUnit runners), the TestRunStarted handler is invoked, then tests are executed, then TestRunFinished handler is invoked, everything is fine
CLI runner executed in IDEA
<img width="774" alt="CLI runner executed in IDEA" src="https://github.com/user-attachments/assets/9cfdb5d3-3138-4120-a286-4cd2df1f8627">

JUnit runner executed via Gradle
<img width="916" alt="JUnit runner executed via Gradle" src="https://github.com/user-attachments/assets/28bc1ffb-fe07-4d27-a4c8-1f99e5622a11">

But when I run the tests in parallel, the handler of TestRunStarted gets invoked right after all tests have been executed. Even though the actual event time stored in event object is correct (see the timestamps in log on the screenshots)
CLI runner executed in IDEA
<img width="811" alt="CLI runner executed in IDEA" src="https://github.com/user-attachments/assets/5224c211-e564-4697-9593-4153da076ee2">

JUnit runner executed via Gradle
<img width="1029" alt="JUnit runner executed via Gradle" src="https://github.com/user-attachments/assets/583f3a85-966f-48f9-bebe-0659953e9e83">

In case of CLI runner I specified `--threads 2` in the program args, in case of JUnit runner I specified `cucumber.execution.parallel.enabled=true` inside the junit-platform.properties file. The threads count doesn't affect anything in the latter case, it is reproduced even with one thread****

### ✅ What did you expect to see?

I expected that during the parallel test execution the `TestRunStarted` handler will be executed once the event happens, before the tests are started. So that I could run some set-up logic inside it. Currently It seems to not be practical

### 📦 Which tool/library version are you using?

org.junit:junit-bom:5.11.0
io.cucumber:cucumber-bom:7.18.1

### 🔬 How could we reproduce it?

Minimal reproducible example forked from the cucumber-java skeleton: [cucumber-java-issue-events](https://github.com/MazurK239/cucumber-java-issue-events)

Run feature file via IntelliJ IDEA (CLI runner will be used) or via Gradle

**Comment by mpkorstanje** [comment]
That is intentional. Most plugins can't handle parallel execution. You'll want to use the `ConcurrentEventListener`.

https://github.com/cucumber/cucumber-jvm/blob/0cdf507e357eda48c7496e63de9588b75134aa72/cucumber-plugin/src/main/java/io/cucumber/plugin/EventListener.java#L9-L15

**Comment by mpkorstanje** [comment]
Btw, for setup you might want to look at the `@BeforeAll` and `@AfterAll` hooks from Cucumber or depending on what you need, those recently added to the JUnit 5 Suite Engine.

**Comment by MazurK239** [comment]
Thank you @mpkorstanje. Works like a charm! sorry that I missed this in the documentation.
Didn't know about `@BeforeAll` and `@AfterAll` either, will definitely give them a try

**Comment by MazurK239** [comment]
@mpkorstanje, since there is yet no documentation regarding the use of BeforeAll/AfterAll could you, please, advise, how to correctly use it in Kotlin? As I see, internally, there is a check that this method is static, but with Kotlin it seems impossible to make this check pass

**Comment by mpkorstanje** [comment]
Docs are here:

https://github.com/cucumber/cucumber-jvm/tree/main/cucumber-java#beforeall--afterall

And for Java to Kotlin interop you'll want to read:

https://kotlinlang.org/docs/java-to-kotlin-interop.html#static-methods

**Comment by MazurK239** [comment]
Many thanks!

**Comment by ejbartelds** [comment]
I saw @MazurK239 's question and @mpkorstanje 's answer today, concerning the use of `BeforeAll/AfterAll` with Cucumber in Kotlin, but In ran into a problem when trying to use those. See screenshots.

It appears Cucumber's method scanning in `io.cucumber.java.MethodScanner::scan` sees both the original companion object method (which is not static) AND the synthesized one (due to `@JvmStatic`) (which has been made static).

As a result, I cannot get this to work in Kotlin...

Any thoughts?


![image](https://github.com/user-attachments/assets/a6a6af31-959f-4aff-a915-c33bf109c704)
![image](https://github.com/user-attachments/assets/1b6f2c42-ad1d-4d6f-b1c0-17b3d3926e81)
![image](https://github.com/user-attachments/assets/fb9b907f-2845-493f-803a-b534aeb04bf2)

**Comment by mpkorstanje** [comment]
@ejbartelds for bugs it's usually better to start a new issue.

**Comment by ejbartelds** [comment]
See https://github.com/cucumber/cucumber-jvm/issues/2934   @mpkorstanje
